components/tidb_query_datatype/src/codec/mysql/vector.rs

```diff
@@ -174,233 +109,211 @@ impl<'a> VectorFloat32Ref<'a> {
         }
-
-        Ok(distance as f64)
     }
 
+    #[inline]
     pub fn cosine_distance(&self, b: VectorFloat32Ref<'a>) -> Result<f64> {
-        self.check_dims(b)?;
-        let mut distance: f32 = 0.0;
-        let mut norma: f32 = 0.0;
-        let mut normb: f32 = 0.0;
-        for i in 0..self.len() {
-            unsafe {
-                distance += self.index_unchecked(i) * b.index_unchecked(i);
-                norma += self.index_unchecked(i) * self.index_unchecked(i);
-                normb += b.index_unchecked(i) * b.index_unchecked(i);
-            }
-        }
-
-        let similarity = (distance as f64) / ((norma as f64) * (normb as f64)).sqrt();
-        if similarity.is_nan() {
-            // Divide by zero
-            return Ok(std::f64::NAN);
+        match f32::cosine(self.data(), b.data()) {
+            Some(cosine_similarity) => Ok(cosine_similarity),
+            None => Err(box_err!("Vectors must be of the same length")),
         }
-        let similarity = similarity.clamp(-1.0, 1.0);
-        Ok(1.0 - similarity)
     }
 
     pub fn l1_distance(&self, b: VectorFloat32Ref<'a>) -> Result<f64> {
         self.check_dims(b)?;
         let mut distance: f32 = 0.0;
         for i in 0..self.len() {
-            let diff = unsafe { self.index_unchecked(i) - b.index_unchecked(i) };
+            let diff = self[i] - b[i];
             distance += diff.abs();
         }
 
         Ok(distance as f64)
     }
 
-    pub fn l2_norm(&self) -> f64 {
+    #[inline]
+    pub fn l2_norm(&self) -> Result<f64> {
         // Note: We align the impl with pgvector: Only l2_norm use double
         // precision during calculation.
-        let mut norm: f64 = 0.0;
-        for i in 0..self.len() {
-            let v = unsafe { self.index_unchecked(i) as f64 };
-            norm += v * v;
+        match f32::dot(self.data(), self.data()) {
+            Some(norm) => Ok(norm.sqrt()),
+            None => Err(box_err!("Vectors must be of the same length")),
         }
+    }
+}
 
-        norm.sqrt()
+pub trait VectorFloat32DatumPayloadChunkEncoder: BufferWriter {
+    fn write_vector_float32_to_chunk_by_datum_payload(&mut self, src_payload: &[u8]) -> Result<()> {
+        // VectorFloat32's chunk format is same as binary format.
+        self.write_bytes(src_payload)?;
+        Ok(())
     }
 }
 
-pub trait VectorFloat32Encoder: NumberEncoder {
-    fn write_vector_float32(&mut self, data: VectorFloat32Ref<'_>) -> Result<()> {
-        if !cfg!(target_endian = "little") {
-            return Err(box_err!("VectorFloat32 only support Little Endian"));
-        }
+impl<T: BufferWriter> VectorFloat32DatumPayloadChunkEncoder for T {}
 
-        self.write_u32_le(data.len() as u32)?;
-        self.write_bytes(data.value)?;
+impl<T: BufferWriter> VectorFloat32Encoder for T {}
 
-        Ok(())
-    }
+/// Misaligned, Ref.
+#[derive(Clone, Copy, Debug)]
+pub struct VectorFloat32RefMisaligned<'a> {
```

Comment:
The term "unaligned" is usually used instead of "misaligned" to describe data that is not aligned.